src/igraph/__init__.py

```diff
@@ -3205,6 +3205,71 @@ def TupleList(
         # Construct the graph
         return cls(n, edge_list, directed, {}, vertex_attributes, edge_attributes)
 
+    #####################################################
+    # Constructor for dict-like representation of graphs
+
+    @classmethod
+    def SequenceDict(
+        cls,
+        edges,
+        directed=False,
+        vertex_name_attr="name",
+    ):
+        """Constructs a graph from a dict-of-sequences representation.
+
+        This function is used to construct a graph from a dictionary of
+        sequences (e.g. of lists). For each key x, its corresponding value is
+        a sequence of multiple objects: for each y, the edge (x,y) will be
+        created in the graph. x and y must be either one of:
+
+        - two integers: the vertices with those ids will be connected
+        - two strings: the vertices with those names will be connected
+
+        If names are used, the order of vertices is not guaranteed, and each
+        vertex will be given the vertex_name_attr attribute.
+        """
+        from collections import defaultdict
+
+        # See if we are using names or integers
+        for item in edges:
+            break
+        if not isinstance(item, (int, str)):
+            raise ValueError("Keys must be integers or strings")
+
+        vertex_attributes = {}
+
+        if isinstance(item, str):
+            n = -1
+            name_map = {}
+            edge_list = []
+            for item, sequence in edges.items():
+                vid = name_map[item] = name_map.get(item, n+1)
```

Comment:
There's also an internal type that generates unique (integer) identifiers: [`UniqueIdGenerator`](https://igraph.org/python/doc/api/igraph.datatypes.UniqueIdGenerator.html). That should simplify it to:

```python
for item, sequence in edges.items():
    for item2 in sequence:
        edge_list.append(id[item], id[item2])
```

Reply:
good point